src/librustc/middle/const_eval.rs

```diff
@@ -275,6 +275,26 @@ pub enum ConstVal {
     Tuple(ast::NodeId),
 }
 
+/// Note that equality for `ConstVal` means that the it is the same
+/// constant, not that the rust values are equal. In particular, `NaN
+/// == NaN`.
+impl PartialEq for ConstVal {
+    #[stable(feature = "rust1", since = "1.0.0")]
+    fn eq(&self, other: &ConstVal) -> bool {
+        match (self, other) {
+            (&Float(v1), &Float(v2)) => (v1.is_nan() && v2.is_nan()) || (v1 == v2),
```

Comment:
Is `v1.is_nan() && v2.is_nan()` really what you mean, as opposed to `transmute<f64, u64>(v1) == transmute<f64, u64>(v2)`?